File: src/main/java/net/hotamachisubaru/digger/Commands.java
package net.hotamachisubaru.digger;

import net.hotamachisubaru.digger.util.SchedulerUtil;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.UUID;

public class Commands implements CommandExecutor {
    private final JavaPlugin plugin;
    private final Map<UUID, Integer> blockCount;

    public Commands(JavaPlugin plugin) {
        this.plugin = plugin;
        this.blockCount = new HashMap<>();
    }

    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        if (!(sender instanceof Player player)) {
            // コンソール等は色付け不要なので普通のメッセージでOK
            sender.sendMessage("このコマンドはプレイヤーからのみ実行できます。");
            return true;
        }

        switch (command.getName().toLowerCase(Locale.ROOT)) {
            case "reload" -> {
                SchedulerUtil.runAtPlayer(player, plugin, () -> handleReloadCommand(player));
                return true;
            }
            case "set" -> {
                SchedulerUtil.runAtPlayer(player, plugin, () -> handleSetCommand(player, args));
                return true;
            }
            default -> {
                return false;
            }
        }
    }

    private void handleSetCommand(Player player, String[] args) {
        if (!player.hasPermission("digger.set")) {
            player.sendMessage(color("§c権限がありません"));
            return;
        }
        if (args.length != 2) {
            player.sendMessage(color("§c正しい使い方: /set <プレイヤー名> <値>"));
            return;
        }
        Player targetPlayer = plugin.getServer().getPlayer(args[0]);
        if (targetPlayer == null) {
            player.sendMessage(color("§c指定されたプレイヤーが見つかりません: " + args[0]));
            return;
        }
        int value;
        try {
            value = Integer.parseInt(args[1]);
            if (value < 0) throw new NumberFormatException();
        } catch (NumberFormatException e) {
            player.sendMessage(color("§c設定する値は正の整数でなければなりません。"));
            return;
        }
        blockCount.put(targetPlayer.getUniqueId(), value);
        player.sendMessage(color("§aプレイヤー " + targetPlayer.getName() + " のブロック数を " + value + " に設定しました。"));
        targetPlayer.sendMessage(color("§aあなたのブロック数が " + value + " に設定されました。"));
    }

    private void handleReloadCommand(Player player) {
        if (!player.hasPermission("digger.reload")) {
            player.sendMessage(color("§cあなたにはこのコマンドを実行する権限がありません。"));
            return;
        }
        plugin.reloadConfig();
        player.sendMessage(color("§aconfig.ymlを再読み込みしました。"));
    }

    private Component color(String message) {
        // §での色コードをAdventure Componentに変換
        return LegacyComponentSerializer.legacySection().deserialize(message);
    }
}

File: src/main/java/net/hotamachisubaru/digger/Commands.kt
package net.hotamachisubaru.digger

import org.bukkit.Bukkit
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import org.bukkit.plugin.java.JavaPlugin
import java.util.*

// Commandsクラスは、BukkitのCommandExecutorを実装しており、カスタムコマンドの処理を行います。
// このクラスはプラグインのインスタンスと、プレイヤーごとのブロックカウントを保持するためのマップを受け取ります。
class Commands(private val plugin: JavaPlugin, private val toolMoney: ToolMoney) : CommandExecutor {
    private val blockCount: MutableMap<UUID, Int> = mutableMapOf()

    // onCommandメソッドは、Bukkitがコマンドを受け取ったときに呼び出されます。
    override fun onCommand(sender: CommandSender, command: Command, label: String, args: Array<String>): Boolean {
        // コマンドの送信者がプレイヤーでない場合は、メッセージを送り終了します。
        if (sender !is Player) {
            sender.sendMessage("§cこのコマンドはプレイヤーからのみ実行できます。")
            return true
        }

        // コマンド名に応じて処理を分岐します。
        return when (command.name.lowercase(Locale.getDefault())) {
            "reload" -> {
                // "reload" コマンドの処理を行います。
                handleReloadCommand(sender)
            }
            "set" -> {
                // "set" コマンドの処理を行います。
                handleSetCommand(sender, args)
            }
            else -> {
                // その他のコマンドは未対応とします。
                false
            }
        }
    }

    // handleReloadCommandメソッドは、設定ファイルを再読み込みするためのコマンドを処理します。
    private fun handleReloadCommand(player: Player): Boolean {
        // プレイヤーが権限を持っているか確認します。
        if (!player.hasPermission("digger.reload")) {
            player.sendMessage("§cあなたにはこのコマンドを実行する権限がありません。")
            return true
        }

        // プラグインの設定ファイルを再読み込みします。
        plugin.reloadConfig()
        player.sendMessage("§aconfig.ymlを再読み込みしました。")
        return true
    }

    // handleSetCommandメソッドは、特定のプレイヤーのスコアを設定するコマンドを処理します。
    private fun handleSetCommand(player: Player, args: Array<String>): Boolean {
        // プレイヤーが権限を持っているか確認します。
        if (!player.hasPermission("digger.set")) {
            player.sendMessage("§cあなたにはこのコマンドを実行する権限がありません。")
            return true
        }

        // 引数が2つであることを確認します。
        if (args.size != 2) {
            player.sendMessage("§c使用法: /set <プレイヤー名> <スコア>")
            return true
        }

        // プレイヤー名を取得し、スコアを整数に変換します。
        val playerName = args[0]
        try {
            val newScore = args[1].toInt()
            // Bukkitを使って対象プレイヤーのUUIDを取得します。
            val targetPlayer = Bukkit.getOfflinePlayer(playerName)
            val targetUUID = targetPlayer.uniqueId

            // スコアをブロックカウントマップに設定します。
            blockCount[targetUUID] = newScore
            player.sendMessage("§a${playerName}のスコアを${newScore}に設定しました。")
        } catch (e: NumberFormatException) {
            // スコアが無効な場合のエラーメッセージを送信します。
            player.sendMessage("§c無効なスコアです。数字を入力してください。")
        }
        return true
    }
}
